polkadot/xcm/xcm-builder/src/barriers.rs

```diff
@@ -502,3 +515,125 @@ impl DenyExecution for DenyReserveTransferToRelayChain {
+					origin, nested_xcm, e
+				);
+			})?;
+
+		// Initialise the recursion count the first time we enter recursive execution.
+		let _ = recursion_count::using_once(&mut 1, || {
+			recursion_count::with(|count| {
+				if *count > xcm_executor::RECURSION_LIMIT {
+					log::error!(
+						target: "xcm::barriers",
+						"Recursion limit exceeded, origin: {:?}, nested_xcm: {:?}, count: {count}",
+						origin, nested_xcm
+					);
+
+					return Err(ProcessMessageError::StackLimitReached);
+				}
+				*count = count.saturating_add(1);
+				Ok(())
+			})
+			// Fallback safety in case of an unexpected failure.
+			.unwrap_or(Ok(()))?;
+
+			// Ensure that we always decrement the counter after processing.
+			sp_core::defer! {
+				recursion_count::with(|count| {
+					*count = count.saturating_sub(1);
+				});
+			}
+
+			// Recursively check the nested XCM instructions.
+			Outer::deny_execution(origin, nested_xcm.inner_mut(), max_weight, properties)
+		})?;
+
+		Ok(Ok(ControlFlow::Continue(())))
+	}
+}
+
+/// Denies execution if the XCM contains restricted **local** instructions, first checking at the
+/// top-level and then **recursively** using `DenyNestedXcmInstructions`.
+///
+/// This barrier only applies to **locally executed** XCM instructions (`SetAppendix`,
+/// `SetErrorHandler`, and `ExecuteWithOrigin`). Remote parts of the XCM are expected to be
+/// validated by receiving chain's barrier.
+///
+/// Note: Ensures that restricted instructions do not execute on the local chain, enforcing stricter
+/// execution policies, while allowing remote chains to enforce their own rules.
+pub struct DenyNestedLocalInstructions<Inner>(PhantomData<Inner>);
+
+impl<Inner: DenyExecution> DenyExecution for DenyNestedLocalInstructions<Inner> {
+	/// Denies execution of restricted local nested XCM instructions.
+	///
+	/// This checks for `SetAppendix`, `SetErrorHandler`, and `ExecuteWithOrigin` instruction
+	/// applying the deny filter **recursively** to any nested XCMs found.
+	fn deny_execution<RuntimeCall>(
+		origin: &Location,
+		instructions: &mut [Instruction<RuntimeCall>],
+		max_weight: Weight,
+		properties: &mut Properties,
+	) -> Result<(), ProcessMessageError> {
+		// First, check if the top-level message should be denied.
+		Inner::deny_execution(origin, instructions, max_weight, properties)
+			.inspect_err(|e| {
+				log::warn!(
+					target: "xcm::barriers",
+					"DenyNestedLocalInstructions::Inner denied execution, origin: {:?}, instructions: {:?}, error: {:?}",
+					origin, instructions, e
+				);
+			})?;
+
+		// If the top-level check passes, check nested instructions recursively.
+		instructions.matcher().match_next_inst_while(
+			|_| true,
+			|inst| match inst {
+				SetAppendix(nested_xcm) |
+				SetErrorHandler(nested_xcm) |
+				ExecuteWithOrigin { xcm: nested_xcm, .. } =>
+					DenyNestedXcmInstructions::<Self, Inner>::deny_recursively::<RuntimeCall>(
```

Comment:
hmm, I think that `DenyNestedXcmInstructions` is kind of unnecessary (and for me little bit hard to follow with this Outer/Inner stuff :) )

`DenyNestedXcmInstructions` is private, so I think it would be enough just to have some private helper function:
```
fn deny_recursively<Deny: DenyExecution>(xcm: Xcm) {

         recursion_count::using_once(
             // TODO: all that recursion stuff

            // here just apply recursively
            Deny::deny_execution(xcm, ...)
         )
}
```
and then use here:

```suggestion
                                        // apply `Self` to the `nested_xcm`
					deny_recursively::<Self>::deny_recursively(nested_xcm, ...)
```


Could this be easier? Could this work?

Reply:
Removed DenyNestedXcmInstructions and simplified the recursion logic.